### ISSUE TITLE
WIP Enabling the ability to cancel orders of rebase tokens

### ISSUE THREAD

**Issue by reborn1002** [UniTradeApp/unitrade-contracts#8]
When elastic supply token balance changes the contract now has a different amount in the contract vs what the user sent in originally, meaning their order will never be executed and they are unable to withdraw their funds at all. Second to that if other users add create a new order after the first order now the first order will be executed and give the initial user the second users funds essentially.

This is a really difficult problem to solve as we can't know which ERC-20 tokens are modified or not or when the rebase events will even occur.

The current "solution" is for users to manually send in the difference of the amount from their original order - what the contract balance currently is, then they can cancel their order.

**Cancellation Fix**
One fix that could be implemented into the contract's cancel method is whenever the balance of the token in the contract doesn't equal to the users input amount process a buy order on Uniswap to make up the difference automatically. A method can be created to estimate an amount in ETH that it would cost to buy this amount from Uniswap and if the user sends in at least this amount of ETH utilize it to purchase the token and send them back their original amount of tokens back plus any additional ETH that wasn't used. Not ideal but at least there is an option for getting funds out easier.

**Comment by lamtst** [comment]
How about giving any token with elastic supply a tag? 
So it could be checked every time a sell order is being placed.
Meaning if token type is **elastic**, the user will be allowed only to sell a **maximum** of 80% of the submitted amount and the other 20% will be used as a safety bucket to pull from in order to encounter for fluctuations caused by a rebase.  
Furthermore, the user can choose between 3 safety buckets types  (20%, 30%, 40%): 
The bigger the bucket is, the less the user have to worry about a sell order not being executed because of a rebase.
This way there's no need to check if whether or not a token has been modified, the only thing to look for will be token type tags ( **isElastic**, **isNotElastic**)

**Comment by lamtst** [comment]
> When elastic supply token balance changes the contract now has a different amount in the contract vs what the user sent in originally, meaning their order will never be executed and they are unable to withdraw their funds at all. Second to that if other users add create a new order after the first order now the first order will be executed and give the initial user the second users funds essentially.
> 
> This is a really difficult problem to solve as we can't know which ERC-20 tokens are modified or not or when the rebase events will even occur.
> 
> The current "solution" is for users to manually send in the difference of the amount from their original order - what the contract balance currently is, then they can cancel their order.
> 
> **Cancellation Fix**
> One fix that could be implemented into the contract's cancel method is whenever the balance of the token in the contract doesn't equal to the users input amount process a buy order on Uniswap to make up the difference automatically. A method can be created to estimate an amount in ETH that it would cost to buy this amount from Uniswap and if the user sends in at least this amount of ETH utilize it to purchase the token and send them back their original amount of tokens back plus any additional ETH that wasn't used. Not ideal but at least there is an option for getting funds out easier.

How about giving any token with elastic supply a tag?
So it could be checked every time a sell order is being placed.
Meaning if token type is **elastic**, the user will be allowed only to sell a **maximum of 80%** of the submitted amount and the other 20% will be used as a safety bucket to pull from in order to encounter for fluctuations caused by a rebase.
Furthermore, the user can choose between 3 safety buckets types (20%, 30%, 40%):
The bigger the bucket is, the less the user have to worry about a sell order not being executed because of a rebase.
This way there's no need to check if whether or not a token has been modified, the only thing to look for will be token type tags ( **isElastic**, **isNotElastic**)

**Comment by lamtst** [comment]
Any time a token is in the low list, it can be automatically tagged as **isElastic**

https://etherscan.io/tokens/label/rebase-token

**Comment by reborn1002** [comment]
@lamtst What is a reliable on-chain trustless way to determine if an ERC-20 contract has an elastic supply? We enable the ability to trade tokens with deflationary supplies (fees on transactions) but determining if a token has an elastic supply mechanism is much more difficult imo

**Comment by lamtst** [comment]
@reborn1002 the alternative right now could be to have in the smart contract an array or hash table (with token symbols and contract addresses as key value pairs) along with a method that update it every time a new token is added to the following etherscan rebase token list (https://etherscan.io/tokens/label/rebase-token)

So when placing a sell order, there could be a conditional statement that first checks if the token is in the isElastic array or hash table.
If true then act accordingly.

**Comment by marcelomorgado** [comment]
And if we implement a `sync()` function that adjusts the `amountInOffered` of all orders according to the new balance of that token?
For instance:
```solidity
function sync(IERC20 token) {
  uint256 before = sumOfAllAmountInOffered(address(token));
  uint256 after = token.balanceOf(address(this));

  if (before == after) {
    return;
  }

  for (uint256 i = 0; i < activeOrders.length; ++i) {
    if (activeOrders[i].tokenIn == address(token) {
      activeOrders[i].amountInOffered = activeOrders[i].amountInOffered.mul(after).div(before);
    }
  }
}
```

I'm wondering what's the best moment to call this function since it can be costly. We can expose it for sure but run it once a day couldn't be enough since we don't control when tokens rebase processes run.
We can call it whenever an operation dealing with a rebase token occurs (e.g.: first thing when calling a `cancel` function for such token), but to do that we would have to keep an on-chain list of rebase tokens, I don't see another way to know if a token has an elastic supply by its address.

**In summary**, we can have a `sync()` function and call it as the first thing inside of the main functions but it has the potential to increase the gas cost of these functions, as a workaround, we can just call it when dealing with tokens with elastic supply based on a managed list of them.

**Note 1**: This approach address another issue: Tokens send to the contract accidentally will be distributed among all active orders instead of being locked forever.
**Note 2**: When implementing this we have to check if we can have some kind of issue due to rounding when calculating new amounts.

WDYT @reborn1002?
